[Feat] add llava-onevision, with support for (1) siglip encoder, (2) qwen2 decoder (3) openai api compatible server.

kcz358 opened the pull request:
This is a working PR and I'm putting here to see the changes and welcome any suggestions.

## Motivation
With this PR, we wish to add support for [LLaVA-OneVision](https://arxiv.org/pdf/2408.04619). A new model that accepts multi-type inputs, including single-image with anyres_max_N and multi-image, and video.

## Modification
1. add support for siglip
2. add support for qwen2 (nothing change from previous llava-next).
3. handle single-image with anyres_max_N (at max 729 * (N + 1) tokens) strategy, multi-image with base token strategy (729 tokens per image), video with bilinear interpolation token strategy (196 tokens per frame).
add openai compatible server for launching a backend service for llava-onevision model. for above mentioned multi-type inputs, we all use the same interface as OpenAI's GPT format as the input format (the image-text can also be interleaved).
Checklist
4. add openai compatible server for launching a backend service for llava-onevision model. for above mentioned multi-type inputs, we all use the same interface as OpenAI's GPT format as the input format (the image-text can also be interleaved).

## Checklist
1. Ensure pre-commit pre-commit run --all-files or other linting tools are used to fix potential lint issues.
2. Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
3. Modify documentation as needed, such as docstrings or example tutorials.

kcz358 committed:
Revise chat-ml in chat template

kcz358 committed:
Bring test multi-images server from onevision

kcz358 committed:
Add jobs.mp4

kcz358 committed:
Add mm_utils for llava onevision

kcz358 committed:
Add llava onevision for tp_worker

kcz358 committed:
Fix image_offset for multi_images

Luodian committed:
🔥 remove(jobs.mp4): delete the jobs.mp4 file from the assets directory
✨ feat(test_multi_image_openai_server.py): add functionality to download jobs.mp4
🔧 fix(test_multi_image_openai_server.py): update video path to use cached location

Luodian committed:
🔥 remove(videos): delete unused video file Q98Z4OTh8RwmDonc.mp4

This commit removes the video file that is no longer needed in the project.
Cleaning up unused assets helps to reduce clutter and improve project maintainability.

Luodian committed:
✨ feat(srt_example_llava_v.py): add functionality to download video file
🔧 fix(test_multi_image_openai_server.py): update video URL to match new location

This commit introduces a new feature in `srt_example_llava_v.py` that
downloads a video file from a specified URL and saves it to the user's
cache directory. This change enhances usability by ensuring that the
required video file is readily available for processing.

Additionally, the test file `test_multi_image_openai_server.py` is
updated to reflect the new URL for the video file, ensuring that tests
point to the correct resource location. This helps maintain the integrity
of the tests and ensures they run successfully with the updated video
source.

Luodian committed:
🐛 fix(llava.py): handle ValueError when image_token_index is not found

This change adds a try-except block around the code that retrieves the
index of `image_token_index` in `input_ids`. If the index is not found,
it defaults to 0 instead of raising an error. This improves the robustness
of the code by ensuring that it can handle cases where the image token
is not present in the input, preventing potential crashes during execution.

Luodian commented:
### SGLang Performance (Image)

| Model                     | LLaVA-N-0.5B | LLaVA-N-7B | LLaVA-N-72B |
|----------------------------|--------------|------------|-------------|
| LMMs-Eval LLaVA_Onevision   | ~5s/it       | ~3s/it     | ~3s/it      |
| LMMs-Eval SRT_API (SGLang)  | ~5s/it       | ~5s/it     | ~1.5s/it    |

**BS=1**  &nbsp;&nbsp;&nbsp;&nbsp; **TP=8**

### SGLang Performance (Video)

| Model                     | LLaVA-N-0.5B | LLaVA-N-7B | LLaVA-N-72B |
|----------------------------|--------------|------------|-------------|
| LMMs-Eval LLaVA_Onevision   | ~4s/it       | ~5s/it     | ~8.5s/it    |
| LMMs-Eval SRT_API (SGLang)  | ~3.5s/it     | ~4s/it     | ~6s/it      |

**BS=1**  &nbsp;&nbsp;&nbsp;&nbsp; **TP=8**

*Note: 0.5B and 7B number of heads can't be divided by 8, so it's served only on single GPU.*

Luodian commented:
### Results

**AI2D**

LLaVA-OV-0.5B (original): 57.1%
LLaVA-OV-0.5B (sglang): 56.6%

**VideoMME (32 frames)**

LLaVA-OV-0.5B (original): 44.0%
LLaVA-OV-0.5B (sglang): -%

kcz358 commented:
### More results

#### MME

LLaVA-OV-0.5B (original): 240/1238
LLaVA-OV-0.5B (srt-api, sglang): 240/1232

Notice : Due to the anyres_max_9 strategy, we might need to set `--chunked-prefill-size 16384` when setting up the server otherwise the token numbers might exceed the default 8192

Luodian committed:
update with correct video file from github content

Luodian commented:
## Test in More Output Tokens Scenario

I tested the `lmms-lab/llava-onevision-qwen2-72b-ov`

> python ./test/srt/test_llava_onevision_openai_server.py

**Single Image**

Prompt Tokens: 7186 (7150 visual + 36 text)
Completion Tokens: 487
Completion Tokens/Sec: 45.02

**Video**

Prompt Tokens: 6298 (6272 visual + 26 text)
Completion Tokens: 224
Completion Tokens/Sec: 26.41

But during the completion, I can see the log shows the `gen throughput (tokens/s): 50-51` whether tested on image and video case.

Luodian committed:
change temp to 0 for consistency.

Luodian committed:
📝 docs(README.md): update usage examples for launching the server

This commit adds new examples for launching the server with
additional model paths, tokenizer paths, and parameters such as
host and tp-size. These updates aim to provide clearer guidance
for users on how to utilize the latest models and configurations,
enhancing the overall documentation and user experience.

Luodian committed:
Update README.md

Luodian commented:
> ### SGLang Performance (Image)
> > on AI2D, 10-20 input text tokens + 6-8k input visual tokens, 2-3 output tokens.
> 
> Model	LLaVA-OneVision-0.5B	LLaVA-OneVision-7B	LLaVA-OneVision-72B
> LMMs-Eval LLaVA_Onevision	~5s/it	~3s/it	~3s/it
> LMMs-Eval SRT-API (SGLang)	~5s/it	~5s/it	~1.5s/it
> **BS=1**      **TP=8**
> 
> ### SGLang Performance (Video)
> > on VideoMME, 30-50 input text tokens + 32 * 196 input visual tokens, 2-3 output tokens.
> 
> Model	LLaVA-OneVision-0.5B	LLaVA-OneVision-7B	LLaVA-OneVision-72B
> LMMs-Eval LLaVA_Onevision	~4s/it	~5s/it	~8.5s/it
> LMMs-Eval SRT-API (SGLang)	~3.5s/it	~4s/it	~6s/it
> **BS=1**      **TP=8**
> 
> _Note: 0.5B and 7B number of heads can't be divided by 8, so it's served only on single GPU._

Note: I think this is not the best way to demonstrate the acceleration of sglang since both benchmarks only need model to decode a few tokens. I will try `llava_bench` to better show the difference.

Luodian committed:
refactor: Improve image processing logic in TokenizerManager

This commit refactors the image processing logic in the TokenizerManager class. The changes include:
- Adding a new method `_process_single_image` to handle single image processing
- Updating the `get_pixel_values` method to use the `_process_single_image` method for single image data
- Simplifying the conditional statements for handling image data types

These changes aim to improve the readability and maintainability of the code related to image processing in the TokenizerManager class.

Luodian committed:
refactor: modify unittest for vision part

Luodian committed:
refactor: Improve image processing logic in TokenizerManager

Luodian commented:
Here's the result in more output tokens scenarios. We can see more speed up now.

| videodc_499 | ~6300 input tokens/~150 output tokens          |
|---------------------------------------|----------|
| **Model**                             | **LLaVA-N-72B** |
| LMMs-Eval LLaVA_Onevision             | ~50s/it  |
| LMMs-Eval SRT_API (SGLang)            | ~25s/it  |


| llava_in_the_wild              | ~8000 input tokens/~100 output tokens        |
|---------------------------------------|----------|
| **Model**                             | **LLaVA-N-72B** |
| LMMs-Eval LLaVA_Onevision             | ~20s/it  |
| LMMs-Eval SRT_API (SGLang)            | ~5s/it   |

**BS=1**           **TP=8**

Luodian commented on python/sglang/srt/managers/tp_worker.py:
```diff
@@ -310,11 +310,16 @@ def handle_generate_request(
             req.pixel_values = recv_req.pixel_values
             if req.pixel_values is not None:
+                image_hash = (
+                    sum(recv_req.image_hash)
```
I am confused how to make it more compatible, since `recv.req.image_hash` is a list now. We can't hash the list I think? Here we write a if/else logic to backward comaptible to previous sitatuions.

Luodian committed:
refactor: update hash logic for handling list of hashs.

merrymercy commented on README.md:
```diff
@@ -229,8 +229,13 @@ python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct
 - LLaVA-NeXT-Video
   - see [examples/usage/llava_video](examples/usage/llava_video)
+- [LLaVA-OneVision](https://arxiv.org/abs/2408.03326)
+  - `python3 -m sglang.launch_server --model-path lmms-lab/llava-onevision-qwen2-72b-ov --tokenizer-path lmms-lab/llavanext-qwen-siglip-tokenizer --port=30000 --host=127.0.0.1 --tp-size=8 --chat-template=chatml-llava --chunked-prefill-size=16384`
```
why do you set `--chunked-prefill-size=16384` to be 16384?

kcz358 replied:
Hi, this is because due to `anyres_max_9` strategy, it is very likely that the image tokens + text tokens might exceed 8192 and causing errors when filling image embeddings to input_ids. In most of the test cases drop `--chunked-prefill-size` might be fine but setting it to a higher value could be safer

merrymercy commented on README.md:
```diff
@@ -229,8 +229,13 @@ python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct
   - `python -m sglang.launch_server --model-path liuhaotian/llava-v1.6-vicuna-7b --tokenizer-path llava-hf/llava-1.5-7b-hf --chat-template vicuna_v1.1 --port 30000`
   - `python -m sglang.launch_server --model-path liuhaotian/llava-v1.6-34b --tokenizer-path liuhaotian/llava-v1.6-34b-tokenizer --port 30000`
+  - `python -m sglang.launch_server --model-path lmms-lab/llama3-llava-next-8b --tokenizer-path lmms-lab/llama3-llava-next-8b-tokenizer --port=30000 --host=127.0.0.1 --tp-size=1 --chat-template=llava_llama_3`
+  - `python -m sglang.launch_server --model-path lmms-lab/llava-next-72b --tokenizer-path lmms-lab/llavanext-qwen-tokenizer --port=30000 --host="127.0.0.1" --tp-size=8`
```
do we miss the `--chat-template` argument here?

merrymercy commented on python/sglang/srt/managers/tokenizer_manager.py:
```diff
@@ -189,26 +224,22 @@ async def _handle_single_request(
-                )
-                if return_logprob and logprob_start_len == -1:
-                    logprob_start_len = len(input_ids) - 1
+        if self.is_generation:
```
i guess you messed up the indention here.

merrymercy commented on test/srt/test_vision_openai_server.py:
```diff
@@ -21,9 +22,11 @@ def setUpClass(cls):
+                "chatml-llava",
                 "--tokenizer-path",
-                "llava-hf/llava-1.5-7b-hf",
+                "lmms-lab/llavanext-qwen-siglip-tokenizer",
+                "--chunked-prefill-size",
+                "16384",
```
These are not needed anymore, right?

merrymercy commented on python/sglang/srt/mm_utils.py:
```diff
@@ -17,6 +17,7 @@
 import ast
```
Add a docstring for this file.
Something like
```
"""Utilities for multi-modal models."""
```
I guess we can also update the source of this file to your newer llava version?

kcz358 committed:
Revise commands for llava-next and llava-onevision

kcz358 committed:
Add docs for mm_utils

kcz358 committed:
Fix indentation messed up in tokenizer_manager

kcz358 commented on README.md:
```diff
@@ -227,8 +227,13 @@ python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct
   - `python -m sglang.launch_server --model-path liuhaotian/llava-v1.6-vicuna-7b --tokenizer-path llava-hf/llava-1.5-7b-hf --chat-template vicuna_v1.1 --port 30000`
   - `python -m sglang.launch_server --model-path liuhaotian/llava-v1.6-34b --tokenizer-path liuhaotian/llava-v1.6-34b-tokenizer --port 30000`
+  - `python -m sglang.launch_server --model-path lmms-lab/llama3-llava-next-8b --tokenizer-path lmms-lab/llama3-llava-next-8b-tokenizer --port=30000 --host=127.0.0.1 --tp-size=1 --chat-template=llava_llama_3`
+  - `python -m sglang.launch_server --model-path lmms-lab/llava-next-72b --tokenizer-path lmms-lab/llavanext-qwen-tokenizer --port=30000 --host="127.0.0.1" --tp-size=8`
```
Hi, I merge the tokenizers for all the models in llava-onevision, llava-next-72b and llava-llama3-8b and revise the commands now. Here are the test results

### llava-llama3-8B
<img width="905" alt="image" src="https://github.com/user-attachments/assets/7791a467-732f-4245-86da-8318be14b63e">


### llava-next-72B
<img width="905" alt="image" src="https://github.com/user-attachments/assets/47ff5f19-efa3-4a4b-a71a-2b43bb2ba70f">


### llava-onevision-72B
<img width="903" alt="image" src="https://github.com/user-attachments/assets/b1f2cc55-e367-4019-abdb-40549210657d">

kcz358 committed:
Fix handling image_gridpinpoints error

kcz358 committed:
Fix the order of the logic when handing pinpoints

kcz358 committed:
Fix vision_openai_server unit test bug

kcz358 committed:
Fix import issue in test_vision_openai_server

kcz358 committed:
Add decord in to dependency

Luodian commented:
@merrymercy Can we get this PR merged now?

merrymercy commented:
@Luodian @kcz358 Great work! It is merged now
